Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+      uint32_t toFlatIndx(uint16_t l, uint32_t j, uint32_t i)
+      {
+        uint32_t iSize, indx;
+        if (l == 0)
+          iSize = dim_;
+        else
+          iSize = wc_;
+        indx = l * dim_ * wc_;
+        indx += j * iSize;
+        indx += i;
+        return indx;
+      }
+      /**
+        * Get negative samples probabilistically using the word frequency
+        * table.  This is not used for now.
+        */
+//      uint32_t getNegativeProb(uint32_t target)
+//      {
+//        uint32_t r;
+//        float prob;
+//        do
+//        {
+//          // Pick a random word
+//          r = rand() % wc_;
+//          // Look up the sampling probability in the negatives table
+//          prob = negTab_[r];
+//          // Skip with probability (1 - prob)
+//          if (rand() / RAND_MAX > prob)
+//            r = target; // Force a skip
+//        } while (r == target); // Don't return the target.
+//        return r;
+//      }
+      /**
+        * Randomly choose a word for negative sampling.  This does not
+        * respect the probability distribution of the words in the corpus,
+        * but is adequate and much more efficient.
+        */ 
+      uint32_t getNegative(uint32_t target)
+      {
+        uint32_t r;
+        do
+        {
+          r = rand() % wc_;
+        } while (r == target);
+        return r;
+      }
+      /**
+        * Calculate the neural network Sigmoid function for forward
+        * propagation.
+        */
+      double sigmoid(double x)
+      {
+        //return x;
+        double y;
+        if (x > 25)
+          y = 1.0;
+        else if (x < -25)
+          y = 0.0;
+        else
+          y = 1.0 / (1.0 + exp(-x));
+        return y;
+      }
+      /** Calculate the loss of a given sample.
+        * Uses linear loss rather than log loss as it
+        * is easier to evaluate and understand.
+        * label is either 1 or 0, and represents to
+        * target value.
+        */
+      float calcLoss(float score, uint32_t label)
+      {
+        float loss;
+        if (label == 1)
+          loss = 1 - score;
+          //loss = -log(score + .00001);
+        else
+          loss = score;
+          //loss = -log((1-score) + .00001);
+        return loss;
+      }
+      /**
+        * Forward and Back Propagation to adjust weights.
+        * Processes a single training sample.
+        * For each sample, processes the positive case
+        * (i.e. the sample) as well as a set of randomly
+        * selected negative samples.  Saves the weights
+        * and the weight updates (deltas).
+        * First, propagate the inputs (the context variables)
+        * to the output to compute the output (score).
+        * Then back-propagate errors (i.e. score - label)
+        * back to the output weights and the input weights.
+        */
+      void fbPropagate(tp& train)
+      {
+        //assert(train.main > 0 & train.main <= wc_);
```

Comment:
in general commented out code should either be removed, or labelled with a comment indicating when it might be useful to enable (e.g. tracing/debugging)

Reply:
Removed